### PR TITLE
Optimize count()

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Choose the smallest column to count if possible.

Category (leave one):
- Performance Improvement

## before
```
[
{
    "hostname": "201.nobida.cn",
    "main_metric": "total_time",
    "num_cores": 20,
    "num_threads": 40,
    "path": "/home/amos/git/chorigin/dbms/tests/performance/count.xml",
    "ram": 134765342720,
    "runs": [
        {
            "bytes_per_second": 6080983856.885921,
            "memory_usage": 0,
            "min_time": 0.012000,
            "quantiles": {
                "0.1": 0.012422,
                "0.2": 0.012602,
                "0.3": 0.012762,
                "0.4": 0.012919,
                "0.5": 0.013076,
                "0.6": 0.013218,
                "0.7": 0.013385,
                "0.8": 0.013602,
                "0.9": 0.013897,
                "0.95": 0.014164,
                "0.99": 0.014653,
                "0.999": 0.021527,
                "0.9999": 0.022115
            },
            "queries_per_second": 75.946258,
            "query": "SELECT count() FROM data",
            "query_index": 0,
            "rows_per_second": 760122982.110740,
            "total_time": 15.142287
        }
    ],
    "server_version": "19.12.1",
    "test_name": "count",
    "time": "2019-07-17 02:23:49"
}
]
```

## after
```
[
{
    "hostname": "201.nobida.cn",
    "main_metric": "total_time",
    "num_cores": 20,
    "num_threads": 40,
    "path": "/home/amos/git/chorigin/dbms/tests/performance/count.xml",
    "ram": 134765342720,
    "runs": [
        {
            "bytes_per_second": 27519267822.779499,
            "memory_usage": 0,
            "min_time": 0.002000,
            "quantiles": {
                "0.1": 0.002741,
                "0.2": 0.002768,
                "0.3": 0.002791,
                "0.4": 0.002812,
                "0.5": 0.002834,
                "0.6": 0.002864,
                "0.7": 0.002903,
                "0.8": 0.002963,
                "0.9": 0.003089,
                "0.95": 0.003214,
                "0.99": 0.003691,
                "0.999": 0.009070,
                "0.9999": 0.011122
            },
            "queries_per_second": 343.893812,
            "query": "SELECT count() FROM data",
            "query_index": 0,
            "rows_per_second": 3439908477.847437,
            "total_time": 10.305507
        }
    ],
    "server_version": "19.12.1",
    "test_name": "count",
    "time": "2019-07-17 02:22:44"
}
]
```

### CODE REVIEW

File: dbms/src/Storages/IStorage.h [hunk 1]
```diff
@@ -82,6 +82,8 @@ class IStorage : public std::enable_shared_from_this<IStorage>
     /// Returns true if the storage supports deduplication of inserted data blocks.
     virtual bool supportsDeduplication() const { return false; }
 
+    using ColumnSizeByName = std::unordered_map<std::string, ColumnSize>;
+    virtual ColumnSizeByName getColumnSizes() const { return {}; }
```
Comment (alexey-milovidov) [inline dbms/src/Storages/IStorage.h]: A comment that it's optional and the result is intended to use for query optimization.

File: dbms/src/Columns/IColumn.h [hunk 2]
```diff
@@ -24,6 +24,20 @@ namespace ErrorCodes
 class Arena;
 class ColumnGathererStream;
 
+struct ColumnSize
```
Comment (alexey-milovidov) [inline dbms/src/Columns/IColumn.h]: It is specific for MergeTree. But we use only uncompressed size.
Maybe it's better to move this struct to Storage.h? (at least).

**Comment by amosbird** [comment]
Heh, just found a relevant issue https://github.com/yandex/ClickHouse/issues/5813  .

File: dbms/src/Interpreters/ExpressionAnalyzer.cpp [hunk 3]
```diff
@@ -1040,7 +1040,28 @@ void ExpressionAnalyzer::collectUsedColumns()
 
     /// You need to read at least one column to find the number of rows.
     if (select_query && required.empty())
-        required.insert(ExpressionActions::getSmallestColumn(source_columns));
+    {
+        /// We will find a column with minimum compressed size. Because it is the column that is cheapest to read.
```
Comment (4ertus2) [inline dbms/src/Interpreters/ExpressionAnalyzer.cpp]: Could we extract this logic into separate function? A good function name is better then the best of the best comment :)

Comment (amosbird) [reply dbms/src/Interpreters/ExpressionAnalyzer.cpp]: Not sure if worth it for now. I guess we can have a dedicated function next time when more sophisticated rules are added.

**Comment by alexey-milovidov** [comment]
Ok.

Now the next step:
We have exact information about the number of rows in MergeTree tables. If the user do simple `SELECT count() ... FROM table` query, we can return the answer instantly.

**Comment by alexey-milovidov** [comment]
BTW, we can do better in this variant too:
- across all columns of minimum compressed size, select a column with minimum size of value in memory:

```
┌─database─┬─table────────────┬─name──────────────────┬─type───────────┬─default_kind─┬─default_expression─┬─data_compressed_bytes─┬─data_uncompressed_bytes─┬─marks_bytes─┬─comment─┬─is_in_partition_key─┬─is_in_sorting_key─┬─is_in_primary_key─┬─is_in_sampling_key─┬─compression_codec─┐
│ default  │ hits_100m_single │ IsEvent               │ UInt8          │              │                    │                445561 │               100000000 │      195328 │         │                   0 │                 0 │                 0 │                  0 │                   │
│ default  │ hits_100m_single │ SocialAction          │ String         │              │                    │                445561 │               100000000 │      195328 │         │                   0 │                 0 │                 0 │                  0 │                   │
│ default  │ hits_100m_single │ SocialNetwork         │ String         │              │                    │                445561 │               100000000 │      195328 │         │                   0 │                 0 │                 0 │                  0 │             
```

In current implementation, ClickHouse decided to read String column, though it's better to read UInt8 column.

**Comment by amosbird** [comment]
> select a column with minimum size of value in memory:

You mean `data_uncompressed_bytes` or the type size? And why does `IsEvent` have the same `data_uncompressed_bytes` as `SocialAction`?

**Comment by alexey-milovidov** [comment]
> You mean data_uncompressed_bytes or the type size?

type size. UInt8 takes the same size on disk than empty String but it is much cheaper to read.

**Comment by alexey-milovidov** [comment]
> And why does IsEvent have the same data_uncompressed_bytes as SocialAction?

All SocialAction are empty strings.

**Comment by shangshujie365** [comment]
> Ok.
> 
> Now the next step:
> We have exact information about the number of rows in MergeTree tables. If the user do simple `SELECT count() ... FROM table` query, we can return the answer instantly.

And we can get 'rows' from system.parts, it can be used for query 'select count() from table where partition = xxx '

**Comment by alexey-milovidov** [comment]
@shangshujie365 True.
(you mean `_part` virtual column)

**Pull request merged**